src/backend/distributed/utils/background_jobs.c

```diff
@@ -0,0 +1,1323 @@
+
+	/* TODO this is here for debugging purposses, remove before merge. */
+	if (BackgroundTaskMonitorDebugDelay)
+	{
+		pg_usleep(30 * 1000 * 1000);
+	}
+
+	MemoryContext perTaskContext = AllocSetContextCreate(CurrentMemoryContext,
+														 "PerTaskContext",
+														 ALLOCSET_DEFAULT_MINSIZE,
+														 ALLOCSET_DEFAULT_INITSIZE,
+														 ALLOCSET_DEFAULT_MAXSIZE);
+
+	/*
+	 * First we find all jobs that are running, we need to check if they are still running
+	 * if not reset their state back to scheduled.
+	 */
+	{
+		StartTransactionCommand();
+		PushActiveSnapshot(GetTransactionSnapshot());
+
+		ResetRunningBackgroundTasks();
+
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+	}
+
+
+	MemoryContext oldContextPerJob = MemoryContextSwitchTo(perTaskContext);
+	TimestampTz backgroundWorkerFailedStartTime = 0;
+	bool hasTasks = true;
+	while (hasTasks)
+	{
+		MemoryContextReset(perTaskContext);
+
+		CHECK_FOR_INTERRUPTS();
+
+		InvalidateMetadataSystemCache();
+		StartTransactionCommand();
+		PushActiveSnapshot(GetTransactionSnapshot());
+
+		/*
+		 * We need to load the task into the perTaskContext as we will switch contexts
+		 * later due to the committing and starting of new transactions
+		 */
+		MemoryContext oldContext = MemoryContextSwitchTo(perTaskContext);
+		BackgroundTask *task = GetRunnableBackgroundTask();
+
+		if (!task)
+		{
+			MemoryContextSwitchTo(oldContext);
+
+			PopActiveSnapshot();
+			CommitTransactionCommand();
+
+			hasTasks = false;
+			break;
+		}
+
+		/* we load the database name and username here as we are still in a transaction */
+		char *databaseName = get_database_name(MyDatabaseId);
+		char *userName = GetUserNameFromId(task->owner, false);
+
+		MemoryContextSwitchTo(oldContext);
+
+		PopActiveSnapshot();
+		CommitTransactionCommand();
+
+		MemoryContextSwitchTo(perTaskContext);
+
+		/*
+		 * The background worker needs to be started outside of the transaction, otherwise
+		 * it will complain about leaking shared memory segments used, among other things,
+		 * to communicate the output of the backend.
+		 */
+		dsm_segment *seg = NULL;
+		BackgroundWorkerHandle *handle =
+			StartCitusBackgroundTaskExecuter(databaseName, userName, task->command,
+											 task->taskid, &seg);
+
+		if (handle == NULL)
+		{
+			/*
+			 * We are unable to start a background worker for the task execution.
+			 * Probably we are out of background workers. Warn once and restart the loop
+			 * after a short sleep.
+			 */
+			if (backgroundWorkerFailedStartTime == 0)
+			{
+				ereport(WARNING, (errmsg("unable to start background worker for "
+										 "background task execution")));
+				backgroundWorkerFailedStartTime = GetCurrentTimestamp();
+			}
+
+			const long delay_ms = 1000;
+			(void) WaitLatch(MyLatch, WL_LATCH_SET | WL_TIMEOUT | WL_EXIT_ON_PM_DEATH,
+							 delay_ms, WAIT_EVENT_PG_SLEEP);
+			ResetLatch(MyLatch);
+
```

Comment:
do need to do anything about the DSM segment in this case?

Reply:
good catch, I thought I added it here already, missed that.

Reply:
actually, you did, I see it's in StartCitusBackgroundTaskExecuter.

Reply:
😱 that would have cause a segfault if we would have cleared it here as well. I was actually contemplating if I should do a `NULL` check on seg before clearing it, but concluded that the dsm functions would have thrown an error if the memory was not allocated.

Completely missed that the return pointer would have not been set.